### PR TITLE
add internal `vmappable` interface (part 1)

### PR DESCRIPTION
**Pull request by mattjj**
This is the first part of a change to add a 'vmappable' typeclass. It's for internal use only at the moment but we may want to open it up to users at some point.

We currently have two ways to add new data types to jax, in the sense of the data types being allowed as inputs and outputs of transformed functions:
1. We can add a new type to jaxprs. That's fully general, and decently modularized, but it is pretty heavyweight: lots of table entries have to be added, reflecting the fact that every part of the system has to learn about the new data type. For example, it needs a lowering to HLO, it needs a tangent type for AD, it needs to be supported by the vmap utility functions, etc.
2. We can add a new pytree type. That's lightweight (no other part of the system has to learn about the new data type!), but pretty restrictive: the data type has to be isomorphic to a tuple (plus some limited metadata). That means we don't get to control how the new data type interacts with, say, AD or vmapping: those interactions are fixed for us (e.g. the tangent type of our new type is determined by its isomorphism to tuples and the tangent type for tuples).

The high level idea here is to add a new mechanism for each transform which gives us both flexibility and light weight.

We want this new mechanism because we want to add some new data types (e.g. PRNGKeyArray, arrays with general element types, sparse array types) which require more general interactions with transformations than afforded by pytrees, but we don't want to plumb them through jaxprs (at least at first). For example, these kinds of mechanisms mean we won't need wrappers for e.g. `grad` and `vmap` in the sparse array code.

This PR starts adding such a mechanism for `vmap`, though we'll also want some others (e.g. for AD). It's just a start because there's immediate follow-up work to do:
* make `vmap` not use pytrees at all, by instead registering all pytree-registered types as vmappable (using the provided isomorphism with tuples to instantiate the vmappable typeclass);
* document the design and the user api;
* add other typeclasses, so that we can e.g. unregister PRNGKeyArray as a pytree.

In other words, I'm attempting some land-and-iterate progress here. This bit should already be useful!

cc @froystig @jakevdp

### CODE REVIEW

**Comment by jakevdp** [comment]
This is great! I'm really looking forward to trying this with the `sparse.BCOO` class. Briefly, the issue there is that sparse array shapes are not related to the size of the underlying dense buffers, but rather are stored in the pytree metadata. What that means is that when you vmap over a sparse array, the shape is not appropriately updated. The current workaround involves only storing a portion of the shape when the object is flattened, and then reconstructing the full shape at unflattening: https://github.com/google/jax/blob/752823e568cd080419d3d6d38ff51ecde16f7421/jax/experimental/sparse/bcoo.py#L1071-L1094

I think this vmappable approach will make things much cleaner.

**Review by froystig — APPROVED** [review]

File: jax/interpreters/batching.py [hunk 1]
```diff
@@ -31,51 +33,71 @@
...
+    x_ = trace.full_raise(x)
+    return matchaxis(trace.axis_name, axis_size, x_.batch_dim, spec, x_.val)
+from_elt_handlers: Dict[Type, FromEltHandler] = {}
+
+def make_iota(axis_size: AxisSize) -> Array:
+  handler = make_iota_handlers.get(type(axis_size))
+  if handler:
+    return handler(axis_size)
+  else:
+    return np.arange(axis_size)  # TODO(mattjj): make this a jax.lax.iota?
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/interpreters/batching.py:71]: A funny thing!

`vmap` now creates this `arange`/`iota`, of size equal to the axis size, every time it's called. Since we don't have any laziness built into lax anymore, unless we add a layer of laziness here (because nothing is _using_ this iota yet, though one of our dynamic shape experiments did), we'll actually end up creating this array then ignoring it. Under a `jit` that'd be DCE'd but outside a `jit` it'll actually get created.

I kept `np.arange` because I think it's the cheaper option; `np.arange(1024)` takes <1µs, whereas `lax.iota(np.int32, 1024)` seems to take ~10x that, and on other backends it'll temporarily use some device memory.

The difference is probably immaterial but I was feeling tentative about it so I decided to go with the safer option at first... I might end up adding in some laziness here if it doesn't make the API too weird.

[the remark this answers is not in the document]
Comment (mattjj) [reply jax/interpreters/batching.py]: I ended up adding some laziness to work around an unrelated issue: jax2tf uses vmap to add an _unbounded-size_ batch axis. We had written this PR assuming there would always be an axis size _bound_ available when calling the vmapped function; after all, such a bound must exist to actually carry out a vmapped computation without requiring infinite memory. But with a staging system that allows sizes (and even their bounds) to be specified later, that information may not be available when running the vmapped computation. And jax2tf is such a staging system.

I consider this laziness to be a temporary workaround. We have a more general mechanism which could be used, namely use of the axis_size primitive, but I didn't switch to using that yet because it requires a name for the mapped axis (and so pulling on that thread might lead to several other changes...).
